core/src/main/java/com/linecorp/armeria/internal/common/TimeoutScheduler.java

```diff
@@ -158,13 +164,54 @@ public boolean isTimedOut() {
         return timeoutController.isTimedOut();
     }
 
+    public CompletableFuture<Void> whenTimingOut() {
+        if (timeoutController == null) {
+            if (timingOutFuture == null) {
+                timingOutFuture = new CompletableFuture<>();
+            }
+            return UnmodifiableFuture.wrap(timingOutFuture);
+        }
+        if (timingOutFuture == null) {
+            return timingOutFuture = UnmodifiableFuture.wrap(timeoutController.whenTimingOut());
+        }
+        if (timingOutFuture instanceof UnmodifiableFuture) {
+            return timingOutFuture;
+        }
+
+        return timingOutFuture = UnmodifiableFuture.wrap(timingOutFuture);
+    }
+
+    public CompletableFuture<Void> whenTimedOut() {
+        if (timeoutController == null) {
+            if (timedOutFuture == null) {
+                timedOutFuture = new CompletableFuture<>();
+            }
+            return UnmodifiableFuture.wrap(timedOutFuture);
+        }
+        if (timedOutFuture == null) {
+            return timedOutFuture = UnmodifiableFuture.wrap(timeoutController.whenTimedOut());
+        }
+        if (timedOutFuture instanceof UnmodifiableFuture) {
+            return timedOutFuture;
+        }
+
+        return timedOutFuture = UnmodifiableFuture.wrap(timedOutFuture);
+    }
+
     public void setTimeoutController(TimeoutController timeoutController, EventLoop eventLoop) {
         requireNonNull(timeoutController, "timeoutController");
         requireNonNull(eventLoop, "eventLoop");
         checkState(this.timeoutController == null, "timeoutController is set already.");
+
+        if (timingOutFuture != null) {
+            timeoutController.whenTimingOut().thenRun(() -> timingOutFuture.complete(null));
+        }
+        if (timedOutFuture != null) {
+            timeoutController.whenTimedOut().thenRun(() -> timedOutFuture.complete(null));
+        }
```

Comment:
Consider thread safety - what happens when one thread calls `setTimeoutController()` and another calls `when*()` methods above?

Reply:
I've tried to use CAS here, but it was hard. Because this callback has to work only if tim*OutFuture was allocated before this block. 

Let me use `synchronized` block.